protocols/bgp/server/bmp_router.go

```diff
@@ -0,0 +1,333 @@
+		}
+	}
+
+	r.logger.Info(logMsg)
+}
+
+func (r *router) processTerminationMsg(msg *bmppkt.TerminationMessage) {
+	const (
+		stringType = 0
+		reasonType = 1
+
+		adminDown     = 0
+		unspecReason  = 1
+		outOfRes      = 2
+		redundantCon  = 3
+		permAdminDown = 4
+	)
+
+	logMsg := fmt.Sprintf("Received termination message from %s: ", r.address.String())
+	for _, tlv := range msg.TLVs {
+		switch tlv.InformationType {
+		case stringType:
+			logMsg += fmt.Sprintf("Message: %q", string(tlv.Information))
+		case reasonType:
+			reason := convert.Uint16b(tlv.Information[:2])
+			switch reason {
+			case adminDown:
+				logMsg += "Session administratively down"
+			case unspecReason:
+				logMsg += "Unespcified reason"
+			case outOfRes:
+				logMsg += "Out of resources"
+			case redundantCon:
+				logMsg += "Redundant connection"
+			case permAdminDown:
+				logMsg += "Session permanently administratively closed"
+			}
+		}
+	}
+
+	r.logger.Warning(logMsg)
+
+	r.con.Close()
+	for n := range r.neighbors {
+		r.peerDown(n)
+	}
+}
+
+func (r *router) processPeerDownNotification(msg *bmppkt.PeerDownNotification) {
+	r.neighborsMu.Lock()
+	defer r.neighborsMu.Unlock()
+
+	if _, ok := r.neighbors[msg.PerPeerHeader.PeerAddress]; !ok {
+		r.logger.Warningf("Received peer down notification for %v: Peer doesn't exist.", msg.PerPeerHeader.PeerAddress)
+		return
+	}
+
+	r.peerDown(msg.PerPeerHeader.PeerAddress)
+}
+
+func (r *router) peerDown(addr [16]byte) {
+	if r.neighbors[addr].fsm != nil {
+		if r.neighbors[addr].fsm.ipv4Unicast != nil {
+			r.neighbors[addr].fsm.ipv4Unicast.bmpDispose()
+		}
+
+		if r.neighbors[addr].fsm.ipv6Unicast != nil {
+			r.neighbors[addr].fsm.ipv6Unicast.bmpDispose()
+		}
+	}
+
+	delete(r.neighbors, addr)
+}
+
+func (r *router) processPeerUpNotification(msg *bmppkt.PeerUpNotification) error {
+	r.neighborsMu.Lock()
+	defer r.neighborsMu.Unlock()
+
+	if _, ok := r.neighbors[msg.PerPeerHeader.PeerAddress]; ok {
+		return fmt.Errorf("Received peer up notification for %v: Peer exists already", msg.PerPeerHeader.PeerAddress)
+	}
+
+	if len(msg.SentOpenMsg) < packet.MinOpenLen {
+		return fmt.Errorf("Received peer up notification for %v: Invalid sent open message: %v", msg.PerPeerHeader.PeerAddress, msg.SentOpenMsg)
+	}
+
+	sentOpen, err := packet.DecodeOpenMsg(bytes.NewBuffer(msg.SentOpenMsg[19:]))
+	if err != nil {
+		return fmt.Errorf("Unable to decode sent open message sent from %v to %v: %v", r.address.String(), msg.PerPeerHeader.PeerAddress, err)
+	}
+
+	if len(msg.ReceivedOpenMsg) < packet.MinOpenLen {
+		return fmt.Errorf("Received peer up notification for %v: Invalid received open message: %v", msg.PerPeerHeader.PeerAddress, msg.ReceivedOpenMsg)
+	}
+
+	recvOpen, err := packet.DecodeOpenMsg(bytes.NewBuffer(msg.ReceivedOpenMsg[19:]))
+	if err != nil {
+		return fmt.Errorf("Unable to decode received open message sent from %v to %v: %v", msg.PerPeerHeader.PeerAddress, r.address.String(), err)
+	}
+
+	addrLen := net.IPv4len
```

Comment:
As discussed, a comment may be good for this :)

Reply:
Fixed.